### PR TITLE
Enable and select fp16 feature

### PR DESCRIPTION
**Pull request by xhcao**
This patch defined select fp16 path according to user's request
and driver support.

DEV

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4319)

### CODE REVIEW

**Comment by xhcao** [comment]
@qjia7 @axinging @haoyunfeix @gyagp Please take a look, thank you.

File: tfjs-backend-webgpu/src/kernels/FromPixels.ts [hunk 1]
```diff
@@ -24,6 +24,35 @@ import * as webgpu_program from './webgpu_program';
 import {FromPixelsProgram} from './FromPixels_utils/from_pixels_webgpu';
 import {util} from '@tensorflow/tfjs-core';
 
+var uint8ToFloat16Table = [
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/FromPixels.ts:27]: Can you add a comment above the line to explain how to get this table? What's your reference or algorithm?

File: tfjs-core/src/io/io_utils.ts [hunk 2]
```diff
@@ -152,16 +154,37 @@ export function decodeWeights(
           new Uint16Array(byteBuffer);
       if (dtype === 'float32') {
         if (quantization.dtype === 'uint8' || quantization.dtype === 'uint16') {
-          values = new Float32Array(quantizedArray.length);
-          for (let i = 0; i < quantizedArray.length; i++) {
-            const v = quantizedArray[i];
-            values[i] = v * quantization.scale + quantization.min;
+          if (env().getBool('FLOAT16') && env().getBool('DRIVER_SUPPORT_FLOAT16'))
```
Comment (qjia7) [inline tfjs-core/src/io/io_utils.ts:157]: Maybe it's better to put the fp16 checking into a helper function.

File: tfjs-core/src/types.ts [hunk 3]
```diff
@@ -85,34 +87,47 @@ export interface RecursiveArray<T extends any> {
...
 
 enum UpcastFloat32AndMap {
   'float32' = 'float32',
+  'float16' = 'float32',
   'int32' = 'float32',
   'bool' = 'float32',
   'complex64' = 'complex64'
 }
 
+enum UpcastFloat16AndMap {
```
Comment (qjia7) [inline tfjs-core/src/types.ts:112]: Do you need some test cases to cover these upcast type?

File: tfjs-core/src/io/io_utils.ts [hunk 4]
```diff
@@ -152,16 +154,37 @@ export function decodeWeights(
...
+          {
+            const floatVaules = new Float32Array(quantizedArray.length);
```
[the remark this answers is not in the document]
Comment (xhcao) [reply tfjs-core/src/io/io_utils.ts:159]: Done

File: tfjs-core/src/globals.ts [hunk 5]
```diff
@@ -52,6 +52,18 @@ export function enableDebugMode(): void {
...
 
+/**
+ * Enables fp16 mode which allows developers to use 16bit float arithmetic
+ * feature and 16bit storage features for improving performance.
+ *
+ * The fp16 mode is available on webgpu backend.
+ *
+ * @doc {heading: 'Environment'}
+ */
+export function enableFp16Mode(): void {
```
[the remark this answers is not in the document]
Comment (xhcao) [reply tfjs-core/src/globals.ts:63]: It should be used by users, it should be a tfjs api. Users should explicitly enable fp16 feature in their demos.

File: tfjs-backend-webgpu/src/kernels/FromPixels.ts (hunk 1 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/kernels/FromPixels.ts:27]: Done

File: tfjs-core/src/io/io_utils.ts (hunk 2 above)
Comment (xhcao) [reply tfjs-core/src/io/io_utils.ts:157]: Let's keep it firstly.

File: tfjs-core/src/tensor_util.ts [hunk 6]
```diff
@@ -21,7 +21,7 @@ import {upcastType} from './types';
 import {assert} from './util';
 
 export function makeTypesMatch<T extends Tensor>(a: T, b: T): [T, T] {
-  if (a.dtype === b.dtype) {
+  if ((a.dtype === b.dtype) || (a.dtype === 'float16' || b.dtype === 'float16')) {
```
[the remark this answers is not in the document]
Comment (xhcao) [reply tfjs-core/src/tensor_util.ts:24]: For example, If one type is float16 and the other one is float. We do not want to upcast these types, we handle them on GPU side.

**Comment by xhcao** [comment]
I had worked around the compiling error that there is not shader-float16 GPUExtensionName in webgpu/types
And addressed Jiajia's comments. Please take a look again, thank you.

**Comment by qjia7** [comment]
Thanks xinghua. Let's see owners' feedback on this. +@pyu10055 @annxingyuan @tafsiri review.

**Comment by xhcao** [comment]
tfjs-backend-webgpu has been de-attached from tfjs-core, this CL cannot compile successfully. I also thought this CL does not introduce float16 feature correctly. I will close it and introduce float16 feature in other CLs, this first CL is https://github.com/tensorflow/tfjs/pull/4445

**Pull request closed without merging**